klippy/extras/sos_filter.py

```diff
@@ -0,0 +1,225 @@
+                    fixed_coeff = to_fixed_32(coeff, self._coeff_int_bits)
+                    fixed_section.append(fixed_coeff)
+                elif coeff != 1.0:  # double check colum 3 is always 1.0
+                    raise ValueError("Coefficient 3 is expected to be 1.0"
+                                     " but was %f" % (coeff,))
+            sos_fixed.append(fixed_section)
+        return sos_fixed
+
+    # convert the SOS filter state matrix (zi) to fixed point format
+    def _convert_state(self, filter_state):
+        sos_state = []
+        for section in filter_state:
+            nun_states = len(section)
+            if nun_states != 2:
+                raise ValueError(
+                    "The number of state elements is %i, must be 2"
+                    % (nun_states,))
+            fixed_state = []
+            for col, value in enumerate(section):
+                fixed_state.append(to_fixed_32(value, self._value_int_bits))
+            sos_state.append(fixed_state)
+        return sos_state
+
+
+# Control an `sos_filter` object on the MCU
+class SosFilter:
+    # fixed_point_filter should be an FixedPointSosFilter instance. A filter of
+    # size 0 will create a passthrough filter.
+    # max_sections should be the largest number of sections you expect
+    # to use at runtime. The default is the size of the fixed_point_filter.
+    def __init__(self, mcu, fixed_point_filter, max_sections=None):
+        self._mcu = mcu
+        self._oid = self._mcu.create_oid()
+        self._filter = fixed_point_filter
+        self._max_sections = max_sections
+        if self._max_sections is None:
+            self._max_sections = self._filter.get_num_sections()
+        self.cmd_set_section = (
+            "sos_filter_set_section oid=%d section_idx=%d"
+                " sos0=%i sos1=%i sos2=%i sos3=%i sos4=%i",
+            "sos_filter_set_section oid=%c section_idx=%c"
+                " sos0=%i sos1=%i sos2=%i sos3=%i sos4=%i")
+        self.cmd_config_state = (
+            "sos_filter_set_state oid=%d section_idx=%d state0=%i state1=%i",
+            "sos_filter_set_state oid=%c section_idx=%c state0=%i state1=%i")
+        self.cmd_activate = (
+            "sos_filter_set_active oid=%d n_sections=%d coeff_int_bits=%d",
+            "sos_filter_set_active oid=%c n_sections=%c coeff_int_bits=%c")
+
+    def get_oid(self):
+        return self._oid
+
+    # create an uninitialized filter object
+    def create_filter(self):
+        self._mcu.add_config_cmd("config_sos_filter oid=%d max_sections=%u"
+            % (self._oid, self._max_sections))
+        self._configure_filter(is_init=True)
+
+    # either setup an init command or send the command based on a flag
+    def _cmd(self, command, args, is_init=False, cq=None):
+        if is_init:
+            self._mcu.add_config_cmd(command[0] % args, is_init=True)
+        else:
+            cmd = self._mcu.lookup_command(command[1], cq=cq)
+            cmd.send(args)
```

Comment:
It is not valid to call `mcu.lookup_command()` after the "mcu connect phase" has completed.  (Otherwise, an mcu/host protocol mismatch wont be reported correctly, and the user will experience cryptic host shutdowns.)

I think it would be preferable to use the common `mcu.add_config_cmd()`, `mcu.register_config_callback()`, and then `mcu.lookup_command()` coding pattern.

Reply:
ok, I'll lookup commands in a config_callback.